### ISSUE TITLE
[question] From SfM with 100% camera positions recovery to a finer detailed structure

### ISSUE THREAD

**Issue by skinkie** [alicevision/meshroom#918]
**Describe the problem**
I am aware of the Depthmap approach but I would like to know if the Pierre Moulon way of creating finer details models using the extraction of more features matching is something that can be achieved within Meshroom, without bruteforcing it. I am looking for a naive approach while I understand that by doing a selection of current features the area on which the features could be extracted per photo could also be _significantly_ reduced. 

**Screenshots**
![image](https://user-images.githubusercontent.com/502394/82907478-f5f24f80-9f66-11ea-8351-7ac2f358742c.png)

**Additional context**
1. Given 110 camera positions, I expect meshroom has the information which imagepairs should be checked, because there are feature to match.
2. Extracting this information should reduce the number of features that should be cross checked just to the previously known "good matches".
3. The basic pain starts now; I think it should be possible to go from the existing CameraInit, to a new FeatureExtraction with a Describer Preset that is significantly higher, or has more Describer Types, directly to FeatureMatching (without ImageMatching) and feeding the ImagePairsList from the last SfM step.

@natowi is there currently any block in the pipeline that would allow to extract the imagepairslist from a succesful SfM?

**Comment by skinkie** [comment]
I am also curious about the following thing. The current imagePairsList consists of a format like:
```
W X Y Z
X Y Z
Y Z
```
How should this file format be read? Is it something like "Any features within W could appear in X, Y, Z"?

**Comment by simogasp** [comment]
No it's not related to the features, it's related to the images id and it means which images must be matched with. In your example, W will be matched with X, Y, Z, then X with Y and Z and so on.
This is used to match the images, that does not mean that image W matches with X,Y,Z

**Comment by simogasp** [comment]
For the first question you are looking for computeStructureFromKnonwPoses, which is currently missing in Meshroom but it is in AliceVision. That takes a reconstructed sfmdata file and recompute the 3D points (+ an option to re-do all the matching using the known geometry)
https://github.com/alicevision/AliceVision/blob/develop/src/software/pipeline/main_computeStructureFromKnownPoses.cpp

**Comment by natowi** [comment]
At the moment computeStructureFromKnonwPoses can be used to re-use the previously reconstructed camera positions from a fixed rig. But results with calculated (external) camera positions were bad (high accuracy is required, which is really difficult to achieve).
Here is a draft on this topic: https://github.com/alicevision/meshroom/wiki/Using-known-camera-positions

**Comment by skinkie** [comment]
> No it's not related to the features, it's related to the images id and it means which images must be matched with. In your example, W will be matched with X, Y, Z, then X with Y and Z and so on.
> This is used to match the images, that does not mean that image W matches with X,Y,Z

@simogasp  This is something I don't get. Could you please elaborate a bit further? My main question: If W, will be matched with X,Y,Z why would there be a line for X matching with Y,Z isn't that implicit?

I created the following python script that can parse the sfm-output from _ConvertSfMFormation_. I would have expected that any featureId result into a more list of observationIds. Using this method I don't get the same results, any feedback is appreciated. I want to understand this direction before I go for the _known poses_, especially because I want to implement a manual matching method ;) 

![image](https://user-images.githubusercontent.com/502394/82938529-459a4080-9f92-11ea-8d7e-38ad6b19ff3a.png)


```python3
#!/usr/bin/env python3

import json
import sys

sfm = json.load(open(sys.argv[1]))

features = {}

for s in sfm['structure']:
    for o in s['observations']:
        os = o["observationId"]
        if o['featureId'] not in features:
            features[o['featureId']] = set([os])
        else:
            features[o['featureId']].add(os)

features2 = {}

for x in features.values():
    y = list(x)
    y.sort(reverse = True)

    if y[0] in features2:
        features2[y[0]].update(set(y))
    else:
        features2[y[0]] = set(y)

for x in features2.values():
    y = list(x)
    y.sort(reverse = True)
    print(' '.join(y))
```

**Comment by natowi** [comment]
> This is something I don't get.

I think this is easier to visualize:

![match](https://user-images.githubusercontent.com/8138537/82941753-31a50d80-9f97-11ea-8786-e0b9d4f0565c.JPG)

**Comment by simogasp** [comment]
> @simogasp This is something I don't get. Could you please elaborate a bit further? My main question: If W, will be matched with X,Y,Z why would there be a line for X matching with Y,Z isn't that implicit?

When you first match all the images if u use a brute force approach you match every image with any other image on the dataset. Obviously, if you already matched X with Y you don't match Y with X.
That's what represents imagePair, i.e. the "planning of how to match the images among them". For example If you were to use the ImageMatching with the vocabulary tree instead of brute force matching every image with all the other, you only match images that are visually similarly.

As for your question, even if W matches with X Y Z, you still have to match X with Y and Z because you don't want to miss matches that could be only between X and Y and not in W and Y. This is more expensive of course but gives more robustness.
If you have already the poses of the cameras you can opt for a faster method like yours but still you risk to miss features

**Comment by skinkie** [comment]
With the help of @Baasje85 I think I can define my problem in reasoning. Given that I have a per feature collection of observations, my assumption was that the first observation with a list of other observation of that same feature would hold. This is false. It misses out on the inner product of these observations. Hence, for every other observation of that same feature, the correlation should also be found between that observation. Now given that this only entails one single feature, the operation is likely to grow closely to brute force matching if the observations have a high correlation.

Now that I have this script, is is probably _cool_ to see if a selection either defined in features or feature-quality defined for example in count, around median + standard deviation could define something like goodness. Now why all these experiments? I want to understand the nobs I can turn. And what would help to reduce the input, when not sacrifice the output. I would like to share an interesting result.

If we would state that the brute force method would be the _maximum_ to achieve because we would check all options. Why would my generated imageMatches perform (worse in time, but that might be a skewed benchmark with Chrome) but better in SfM?

![image](https://user-images.githubusercontent.com/502394/82945243-de35be00-9f9c-11ea-997a-df5284a6288a.png)

```python3
 #!/usr/bin/env python3

import json
import sys

sfm = json.load(open(sys.argv[1]))

features = {}

for s in sfm['structure']:
    for o in s['observations']:
        os = int(o["observationId"])
        if o['featureId'] not in features:
            features[o['featureId']] = set([os])
        else:
            features[o['featureId']].add(os)

features2 = {}

for x in features.values():
    y = list(x)
    y.sort(reverse = True)

    if y[0] in features2:
        features2[y[0]].update(set(y))
    else:
        features2[y[0]] = set(y)

for x in list(features2.values()):
    y = list(x)
    y.sort(reverse = True)

    for i in range(1, len(y) - 1):
        if y[i] in features2:
            features2[y[i]].update(set(y[i:]))
        else:
            features2[y[i]] = set(y[i:])

for x in features2.values():
    y = list(x)
    y.sort(reverse = True)
    print(' '.join([str(z) for z in y]))
```

**Comment by skinkie** [comment]
I would like to share another result. I obviously wanted to try the suggestion of @simogasp by using the output of the SfM step, and feeding this into a pipe where I used the "ultra" setting. The results were poor. While all camera's were located, only 2550 points were found with the default settings. Hence throwing in many more features while knowing the camera poses will not by definition return in a great success. This includes the "you do not know what you are doing" factor. 

@natowi it might be smart to describe the implication of choices in at the start of the pipeline, should influence the latter blocks, unless we have blocks autoconfiguring theirselves. For example if I have used the ultra setting in feature extraction, it might be required to increase iterations at some other points as well.
